packages/frontend/workspace/src/affine/batch-sync-sender.ts

```diff
@@ -0,0 +1,104 @@
+interface SyncUpdateSender {
+  (
+    guid: string,
+    updates: Uint8Array[]
+  ): Promise<{
+    accepted: boolean;
+    retry: boolean;
+  }>;
+}
+
+/**
+ * BatchSyncSender is simple wrapper with vanilla update sync with several advanced features:
+ *   - ACK mechanism, send updates sequentially with previous sync request correctly responds with ACK
+ *   - batching updates, when waiting for previous ACK, new updates will be buffered and sent in single sync request
+ *   - retryable, allow retry when previous sync request failed but with retry flag been set to true
+ */
+export class BatchSyncSender {
+  private buffered = new Map<string, Uint8Array[]>();
+  private job: Promise<void> | null = null;
+  private started = false;
+
+  constructor(private readonly rawSender: SyncUpdateSender) {}
+
+  send(guid: string, update: Uint8Array) {
+    let buffer = this.buffered.get(guid);
+
+    if (!buffer) {
+      buffer = [];
+      this.buffered.set(guid, buffer);
+    }
+
+    buffer.push(update);
+
+    // start the poll right after the first update pushed if not stopped
+    if (this.started && !this.job) {
+      this.next();
+    }
+
+    return Promise.resolve();
+  }
+
+  stop() {
+    this.started = false;
+    this.job = null;
+  }
+
+  start() {
+    this.started = true;
+    this.next();
+  }
+
+  private head(): [string, Uint8Array[]] | undefined {
+    for (const pending of this.buffered.entries()) {
+      if (pending[1].length) {
+        return pending;
+      }
+    }
```

Comment:
If there are multiple subdoc updates in `.buffered`, it seems that the syncing jobs are serialized. do you think maybe can be parallelized to increase throughput?

Reply:
that makes sense

Reply:
done